### PR TITLE
Ainstein LR-D1 driver - RangeFinder

### PR DESCRIPTION
**Pull request by jkronk**
Added support for the [Ainstein LR-D1 altimeter](https://ainstein.ai/lr-d1-uav-long-range-radar-altimeter/) using [@peterbarker original PR as baseline.](https://github.com/ArduPilot/ardupilot/pull/24892)

- This PR supports the newest FW version 18.x, which rearranged data words 5&6
- Updated get_signal_quality_pct, was unable to compile using the previous implementation.
- Added additional SNR checking for 'invalid altitude readings' according to the datasheet.
- Corrected malfunction codes and added reporting for 'Stop use and check' alert codes.
- This has only been tested on bench top, still awaiting flight testing.

Will be flight testing in the next few days.

[LR-D1 Technical Manual](https://ainstein.ai/wp-content/uploads/LR-D1-Technical-User-Manual.docx.pdf)

### CODE REVIEW

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define ALLOW_DOUBLE_MATH_FUNCTIONS
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:16]: this doesn't belong here, remove it

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+#if AP_RANGEFINDER_AINSTEIN_LR_D1_ENABLED
+
+#include "AP_RangeFinder_Ainstein_LR_D1.h"
+
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <GCS_MAVLink/GCS.h>
+
+static const uint8_t PACKET_HEADER_MSB = 0xEB;
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:28]: move these to a constexpr in the header class

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+uint8_t AP_RangeFinder_Ainstein_LR_D1::LRD1Union::calculate_checksum() const
+{
+    uint32_t sum = 0;
+    for (uint8_t i=3; i<31; i++) {
+        sum += buffer[i];
+    }
+    return sum & 0xff;
+}
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:55]: no need to reinvent the wheel with redundant code, please use `crc_sum_of_bytes(&buffer[3], 31-3)`

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    signal_quality_pct = RangeFinder::SIGNAL_QUALITY_MAX;
+
+    check_for_malfunction();
+
+    return true;
+}
+
+int8_t AP_RangeFinder_Ainstein_LR_D1::get_signal_quality_pct() const
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:139]: Simple accessors like this should go into the header. I like that it's const!

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+{
+    return signal_quality_pct;
+}
+
+void AP_RangeFinder_Ainstein_LR_D1::check_for_malfunction() {
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:144]: if you pass `u.packet.malfunction_alert` into it as a variable you can make this a static function. Reducing the scope and not depending on magical member classes is usually a good idea

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    uint8_t malfunction_alert = u.packet.malfunction_alert;
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:145]: if you stored the value as `MalfunctionAlert` then you wouldn't have to do so much casting later on

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    bool move_signature_in_buffer(uint8_t start);
+
+    void check_for_malfunction();
+
+    enum class MalfunctionAlert {
+        Temperature = 0x01,
+        Voltage = 0x02,
+        IFSignalSaturation = 0x40,
+        AltitudeReading = 0x80,
+    };
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:61]: If an enum is actually s bitmask, better to make to more obvious by using (1U<<0), (1U<<1), (1U<<2)... instead of 0x1 0x2, 0x4.. and then have them all lined up i with proper whitespace makes it very obvious at a glance that it's a bitwise mask.

Also, adding ` : uint8_t` to the end of the name might make casting easier

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            uint16_t object1_velocity;
+            uint8_t unused[20];
+            uint8_t checksum;  // (data4+data5+…+data29+data31) bitwise-AND with 0xFF
+        } packet;
+        uint8_t buffer[64];  // each packet is 30 bytes long
+
+        // return checksum calculated from data in buffer
+        uint8_t calculate_checksum() const;
+    } u;
+    uint8_t buffer_used;
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:82]: "buffer_index" might be a better name here

File: libraries/SITL/SIM_Aircraft.cpp [hunk 9]
```diff
@@ -554,7 +554,7 @@ float Aircraft::rangefinder_range() const
 
     // Add some noise on reading
     altitude += sitl->sonar_noise * rand_float();
-
+
```
Comment (magicrub) [inline libraries/SITL/SIM_Aircraft.cpp:557]: remove whitespace

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    buffer[23] = 0xff;
+    buffer[24] = 0xff;
+    buffer[25] = 0xff;
+    buffer[26] = 0xff;
+    buffer[27] = 0xff;
+    buffer[28] = 0xff;
+    buffer[29] = 0xff;
+    buffer[30] = 0xff;
+
+    uint32_t sum = 0;
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp:78]: please re-use the checksums from crc.h

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    buffer[21] = 0xff;
+    buffer[22] = 0xff;
+    buffer[23] = 0xff;
+    buffer[24] = 0xff;
+    buffer[25] = 0xff;
+    buffer[26] = 0xff;
+    buffer[27] = 0xff;
+    buffer[28] = 0xff;
+    buffer[29] = 0xff;
+    buffer[30] = 0xff;
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp:76]: Would it make more sense to memset these?

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  Simulator for the Ainstein LR-D1 Serial rangefinder
+*/
+
+#include "SIM_RF_Ainstein_LR_D1.h"
+
+#include <stdio.h>
+#include <string.h>
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp:22]: what are these includes for?

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        return 0;
+    }
+
+    uint8_t malfunction_alert = 0;
+
+    uint8_t snr;
+    if (alt_cm > 10000) {
+        // out of range @100m
+        snr = 0;
+        alt_cm = 146;
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp:40]: what is this magic number of 146?

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.h [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  Simulator for the Ainstein LR-D1 serial rangefinder
+
+./Tools/autotest/sim_vehicle.py --gdb --debug -v ArduCopter -A --uartF=sim:ainsteinlrd1 --speedup=1
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.h:18]: Usage of --uartX is deprecated

```suggestion
./Tools/autotest/sim_vehicle.py --gdb --debug -v ArduCopter -A --serial5=sim:ainsteinlrd1 --speedup=1
```

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.h [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+param set RNGFND1_TYPE 42
+graph RANGEFINDER.distance
+graph GLOBAL_POSITION_INT.relative_alt/1000-RANGEFINDER.distance
+reboot
+
+arm throttle
+rc 3 1600
+
+# to pass canned data through the simulator:
+/Tools/autotest/sim_vehicle.py --gdb --debug -v ArduCopter -A --uartF=file:libraries/AP_RangeFinder/examples/captures/ainstein_lr-d1.bin --speedup=1
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.h:30]: Usage of --uartX is deprecated
```suggestion
/Tools/autotest/sim_vehicle.py --gdb --debug -v ArduCopter -A --serial5=file:libraries/AP_RangeFinder/examples/captures/ainstein_lr-d1.bin --speedup=1
```

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+            uint8_t length;  // "fixed as 28 bytes"
+            uint8_t malfunction_alert;
+            uint8_t objects_number;  // "fixed as 1 (in version FW 6.0.7.x only), no longer reliable for data validation checks"
+            uint16_t object1_alt;
+            uint8_t object1_snr;
+            uint16_t object1_velocity;
+            uint8_t unused[20];
+            uint8_t checksum;  // (data4+data5+…+data29+data31) bitwise-AND with 0xFF
+        } packet;
+        uint8_t buffer[64];  // each packet is 30 bytes long
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:77]: Why are your union sizes not matching? Code says [64] and comment says 30 but I count 32

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+        // return checksum calculated from data in buffer
+        uint8_t calculate_checksum() const;
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:80]: remove, use functions from crc.h

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+            reading_m = MAX(656, max_distance_cm() * 0.01 + 1);
+            return true;
+        }
+        
+        return false;
+    }
+
+    signal_quality_pct = RangeFinder::SIGNAL_QUALITY_MAX;
+
+    check_for_malfunction();
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:134]: The way this is written not, this has the potential of spamming the GCS. Needs a throttle of some sort to not spam any faster than 1Hz or just once (for each malfunction) by caching the value and only printing on malfunction change.. Don't forget to nerf signal_quality_pct when it's in an error state

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+    */
+#ifndef HAL_BUILD_AP_PERIPH   // we may need to move this elsewhere to the entire function or driver?
+    const auto &ahrs = AP::ahrs();
+    if (fabs(ahrs.roll) > radians(43)) {
+        return false;
+    }
+    if (fabs(ahrs.pitch) > radians(30)) {
+        return false;
+    }
+#endif
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:110]: This should be AP_AHRS_ENABLED instead of a Periph check but really from a driver's perspective this chunk should be removed. There's no guarantee this is used to look straight down at the ground, it could be forward/side/up or whatever orientation. That can be checked via `enum Rotation` The datasheet just says that for proper operation it expects to bounce off a roughly perpendicular surface.. but that's not the driver's responsibility to enforce.

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+// get_reading - read a value from the sensor
+bool AP_RangeFinder_Ainstein_LR_D1::get_reading(float &reading_m)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    const uint8_t num_read = uart->read(&u.buffer[buffer_used], ARRAY_SIZE(u.buffer)-buffer_used);
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:64]: ok, so you're reading up to a max of 2ish backets but never looking beyond the first one in the same time horizon, you look at it the next time get_reading is called? I would think you'd always want to look at the most recent one? This scheme is odd.

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+
+    // sanity checks; see data sheet on these fixed values.
+    if (u.packet.header_lsb != PACKET_HEADER_LSB ||
+        u.packet.device_id != 0x00 ||
+        u.packet.length != 28 ||
+        u.calculate_checksum() != u.packet.checksum) {
+        // sanity checks failed - discard and try again next time we're called:
+        move_signature_in_buffer(1);
+        return false;
+    }
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:88]: You spend the time/code to look at these and store them and potentially check them multiple times. This whole buffer manager parser scheme is bizarre... Especialyl with the moving signature thing. Makes it hard to see whats going on.

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+#include <AP_HAL/utility/sparse-endian.h>
+#include <GCS_MAVLink/GCS.h>
+
+static const uint8_t PACKET_HEADER_MSB = 0xEB;
+static const uint8_t PACKET_HEADER_LSB = 0x90;
+
+// make sure we know what size the packet object is:
+// assert_storage_size<AP_RangeFinder_Ainstein_LR_D1::LRD1Union::LRD1Packet, 32> _assert_storage_lrd1_packet;
+
+bool AP_RangeFinder_Ainstein_LR_D1::move_signature_in_buffer(uint8_t start)
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:34]: not a fan of this method at all

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    const uint8_t num_read = uart->read(&u.buffer[buffer_used], ARRAY_SIZE(u.buffer)-buffer_used);
+    buffer_used += num_read;
+
+    if (buffer_used == 0) {
+        return false;
+    }
+
+    if (u.packet.header_msb != PACKET_HEADER_MSB &&
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:71]: Can I help you with this parser by writing one up for you that is:
- easier to read
- faster
- lower RAM footprint
- lower Flash footprint

File: libraries/AP_HAL_SITL/SITL_State.cpp [hunk 24]
```diff
@@ -599,6 +600,14 @@ void SITL_State::set_height_agl(void)
         if (_terrain != nullptr &&
             _terrain->height_amsl(location, terrain_height_amsl, false)) {
             _sitl->state.height_agl = _sitl->state.altitude - terrain_height_amsl;
+            if (_sitl->state.height_agl < 0) {
+                static bool sent_warning_clamp;
```
Comment (magicrub) [inline libraries/AP_HAL_SITL/SITL_State.cpp:604]:
```suggestion
                static bool sent_warning_clamp = false;
```

File: libraries/AP_HAL_SITL/SITL_State.h [hunk 25]
```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include "AP_HAL_SITL_config.h"
```
Comment (magicrub) [inline libraries/AP_HAL_SITL/SITL_State.h:3]: is this include necessary? Other drivers don't need it

**Review by magicrub — CHANGES_REQUESTED** [review]

**Comment by magicrub** [comment]
I threw this together but haven't run it yet. Let me know what you think
https://github.com/magicrub/ardupilot/commit/3d5e4ff01d6cf1a8a6466c63436f8128b25e736d

**Comment by jkronk** [comment]
Thanks for the feedback, looked at your PR and I am going to use your suggestions. Code is much easier to read. Will do ground tests and check in.

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+            report_malfunction(malfunction_alert);
+        }
+
+        /* From datasheet:
+            Altitude measurements associated with a SNR value 
+            of 13dB or lower are considered erroneous
+        */
+        signal_quality_pct = (snr <= 13 || malfunction_alert != 0) ? RangeFinder::SIGNAL_QUALITY_MIN : RangeFinder::SIGNAL_QUALITY_MAX;
+
+        if (snr == 0) {
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:84]:
```suggestion
        if (snr == 0 || signal_quality_pct != RangeFinder::SIGNAL_QUALITY_MAX) {
```

You also need to set Status::NoData/OutOfRangeHigh/Low/Good around here somewhere

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+    // get a reading
+    bool get_reading(float &reading_m) override;
+
+    // 0 is no return value, 100 is perfect.  false means signal
+    // quality is not available
+    int8_t get_signal_quality_pct() const override { return signal_quality_pct; };
+
+    static void report_malfunction(const uint8_t _malfunction_alert_);
+
+    enum class MalfunctionAlert : uint8_t {
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:52]: Are there other alerts that are being ignored intentionally?

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 28]
```diff
@@ -0,0 +1,119 @@
...
+// get_reading - read a value from the sensor
+bool AP_RangeFinder_Ainstein_LR_D1::get_reading(float &reading_m)
+{
+    if (uart == nullptr || uart->available() == 0) {
+        return false;
+    }
+
+    bool has_data = false;
+
+    uint32_t available = MAX(uart->available(), static_cast<unsigned int>(PACKET_SIZE*4));
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:32]: this cast shouldn't be necessary.. but it's fine

File: libraries/SITL/SIM_Aircraft.cpp [hunk 29]
```diff
@@ -554,11 +554,10 @@ float Aircraft::rangefinder_range() const
 
     // Add some noise on reading
     altitude += sitl->sonar_noise * rand_float();
-
+    
     return altitude;
 }
 
-
```
Comment (magicrub) [inline libraries/SITL/SIM_Aircraft.cpp:561]: remove unnecessary whitespace changes

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+
+    uint8_t malfunction_alert = 0;
+
+    uint8_t snr;
+    if (alt_cm >= 65535) {
+        snr = 0;
+        alt_cm = 65535;
+    } else {
+        snr = 100;
+    }
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp:39]: alt_cm can never be > 0xFFFF so might as well simply this

```suggestion
const uint8_t snr = (alt_cm == 0xFFFF) ? 0 : 100;
```

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+    buffer[3] = 28;  // length
+    buffer[4] = malfunction_alert;
+    buffer[5] = 1;  // object count
+    buffer[6] = alt_cm >> 8;
+    buffer[7] = alt_cm & 0xff;
+    buffer[8] = snr;
+    buffer[9] = 0;  // speed high
+    buffer[10] = 0;  // speed low
+    memset(buffer + 11, 0xff, 20);
+    buffer[31] = crc_sum_of_bytes(&buffer[3], 28);
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp:54]: I'm allergic to magic numbers :)
```suggestion
    buffer[31] = crc_sum_of_bytes(&buffer[3], PACKET_LEN-4); // minus headerMSB, headerLSB, device_id and checksum
```

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp [hunk 32]
```diff
@@ -0,0 +1,57 @@
...
+    buffer[2] = 0;  // device_id
+    buffer[3] = 28;  // length
+    buffer[4] = malfunction_alert;
+    buffer[5] = 1;  // object count
+    buffer[6] = alt_cm >> 8;
+    buffer[7] = alt_cm & 0xff;
+    buffer[8] = snr;
+    buffer[9] = 0;  // speed high
+    buffer[10] = 0;  // speed low
+    memset(buffer + 11, 0xff, 20);
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp:53]: The syntax of `buffer + 11` is fine but you use the other syntax `&buffer[3]` on the line after it. Better to be consistent.
```suggestion
    memset(&buffer[11], 0xff, 20); // unused
```

File: libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp [hunk 33]
```diff
@@ -0,0 +1,57 @@
...
+
+
+    buffer[0] = 0xEB;  // packet header msb
+    buffer[1] = 0x90;  // packet header lsb
+    buffer[2] = 0;  // device_id
+    buffer[3] = 28;  // length
+    buffer[4] = malfunction_alert;
+    buffer[5] = 1;  // object count
+    buffer[6] = alt_cm >> 8;
+    buffer[7] = alt_cm & 0xff;
```
Comment (magicrub) [inline libraries/SITL/SIM_RF_Ainstein_LR_D1.cpp:49]: This is OK but next time I encourage you to use our built-in macros: `LOWBYTE(alt_cm)` and `HIGHBYTE(alt_cm)`

**Review by magicrub — CHANGES_REQUESTED** [review]

**Comment by magicrub** [comment]
still have a few lingering changes (like whitespace stuff) but the commits need to get squashed and the commit text needs to confirm to our standards. Basically, one commit per library

**Comment by jkronk** [comment]
> still have a few lingering changes (like whitespace stuff) but the commits need to get squashed and the commit text needs to confirm to our standards. Basically, one commit per library

Will do, planning on flying this build in the next week or 2. Will update the PR with results.

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h [hunk 34]
```diff
@@ -0,0 +1,60 @@
...
+#include "AP_RangeFinder_config.h"
+
+/*
+ questions:
+ - bug on page 22 on malfunction codes
+ - fixed length seems strange at 28 bytes
+ - definition of snr field is missing in documentation
+ - roll/pitch limits are in conflict, 3.2 vs 
+*/
+
```
Comment (rishabsingh3003) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:12]: There should be a` #if AP_RANGEFINDER_AINSTEIN_LR_D1_ENABLED here`

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h [hunk 35]
```diff
@@ -0,0 +1,60 @@
...
+        return new AP_RangeFinder_Ainstein_LR_D1(_state, _params);
+    }
+
+protected:
+
+    MAV_DISTANCE_SENSOR _get_mav_distance_sensor_type() const override {
+        return MAV_DISTANCE_SENSOR_RADAR;
+    }
+
+    uint32_t initial_baudrate(uint8_t serial_instance) const override {
```
Comment (rishabsingh3003) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:33]: Why is this needed? Can the baudrate of the sensor be changed? If yes, then that can be handled by SERIALx_BAUD

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 36]
```diff
@@ -0,0 +1,115 @@
...
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_RangeFinder_config.h"
+
```
Comment (rishabsingh3003) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:17]: Needs a `#if AP_RANGEFINDER_AINSTEIN_LR_D1_ENABLED`

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h [hunk 37]
```diff
@@ -0,0 +1,60 @@
...
+    static void report_malfunction(const uint8_t _malfunction_alert_);
+
+    enum class MalfunctionAlert : uint8_t {
+        Temperature       = (1U << 0),  // 0x01
+        Voltage           = (1U << 1),  // 0x02
+        IFSignalSaturation= (1U << 6),  // 0x40
+        AltitudeReading   = (1U << 7),  // 0x80
+    };
+
+    static constexpr uint8_t PACKET_SIZE = 32;
```
Comment (rishabsingh3003) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:57]: Our usual way is to use macros, not sure if its a big deal

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 38]
```diff
@@ -0,0 +1,115 @@
...
+        const uint8_t malfunction_alert = buffer[1];
+        reading_m = UINT16_VALUE(buffer[3], buffer[4]) * 0.01;
+        const uint8_t snr = buffer[5];
+
+        has_data = true;
+
+        if (malfunction_alert_prev != malfunction_alert) {
+            malfunction_alert_prev = malfunction_alert;
+            report_malfunction(malfunction_alert);
```
Comment (rishabsingh3003) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:72]: How often can these be sent out? We should probably rate-limit them just in case? 
Should users be allowed to turn this off somehow?

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp [hunk 39]
```diff
@@ -0,0 +1,115 @@
...
+        /* From datasheet:
+            Altitude measurements associated with a SNR value 
+            of 13dB or lower are considered erroneous. 
+
+            SNR values of 0 are considered out of maximum range (655 metres)
+
+            The altitude measurements should not in any circumstances be used as true
+            measurements independently of the corresponding SNR values. 
+        */
+        signal_quality_pct = (snr <= 13 || malfunction_alert != 0) ? RangeFinder::SIGNAL_QUALITY_MIN : RangeFinder::SIGNAL_QUALITY_MAX;
```
Comment (rishabsingh3003) [inline libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:84]: Can the "13" be a macro? 
Our CAN based rangefinders have a SNR parameter. I wonder if we should pull it out and make use of it here.

**Review by rishabsingh3003 — CHANGES_REQUESTED** [review]

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp (hunk 39 above)
Comment (magicrub) [reply libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:84]: I think it's ok to hard code this because the driver is saying the data is bad at this level. The any param could be another level this I've but that could be added to a future PR. Best to get this chunk in before feature creep kills it

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h (hunk 37 above)
Comment (magicrub) [reply libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:57]: I don't think it's a big deal. Does anyone else care?

**Comment by jkronk** [comment]
Looks like the failing tests are due to a 503 error on the CI service.

**Comment by jkronk** [comment]
Thanks @magicrub 

Am flight testing this Friday and will post results then squash my commits

**Comment by jkronk** [comment]
@magicrub @rishabsingh3003 I did notice today the max value cm is an int16 which caps the max value at 32767, this rangefinder technically has a max value of 65535. I was going to override the max_distance_cm() function, but am starting to think it should be changed to a uint16 at the base class level. What are your thoughts on this one?

**Comment by magicrub** [comment]
Is FFFF an error state or is it a valid distance?

**Comment by jkronk** [comment]
> Is FFFF an error state or is it a valid distance?

thats the maximum reading from the altimeter saying its out of range, 500m is what they say is 'reliable max', but correct readings can occur up to 65535

**Comment by magicrub** [comment]
@jkronk well that is definitely a discussion for a PR after this one merges. I see what you mean, int16_t for min_cm and max_cm params. That's unfortunate., However, since valid-ish ranges go all the way up to 65535 then setting the params to 0 is the right thing to do.

How did your testing go? Ready to merge this thing? Just need to clean up the commits and remove the unnecessary whitespace change(s). Let me know if you'd like some help with that.

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h (hunk 34 above)
Comment (magicrub) [reply libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.h:13]: added

File: libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp (hunk 38 above)
Comment (magicrub) [reply libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp:74]: I added a rate limited and a define-out

**Review by magicrub — APPROVED** [review]

**Comment by magicrub** [comment]
I've rebased, re-juggled the commits and addressed some outstanding feedback. I am told that @jkronk did a flight test with this so lets get this in before it bit rots too much
